improve: code block

luanvdw opened the pull request:
Closes: #248
Closes: #249
Closes: #254 

**!NOTE**: This PR will result in breaking changes where the Code components are currently in use. Once approved and merged, I submit my PR for the version bump in cloud with updates to accommodate this rework.

- Refactored Inline code to be its own component.
- Code block now caters for the following additional use cases.
  - Sensitive value with AND without permission to reveal and copy
  - No initial value with AND without permission to reveal and copy
  - Optional masked value for sensitivity.


_*The stories has been updated with these use cases._

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -8,68 +8,129 @@ import React, {
+        ? true
+        : !value && onReveal
+        ? true
+        : false
```
Long `? :` chains like this one are really hard to read / reason about, can we split this up?

madebysid replied:
A boring `if` condition will work here

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -8,68 +8,129 @@ import React, {
-      getSecret?: () => string
+      isSecret?: false
+      permissionToReveal?: never
+      maskedValue?: never
```
Do we need to let people define the masked value from the outside? If not, lets get rid of this prop

luanvdw replied:
No, this is merely optional. It is there to cater when we want a specific value to serve as the masked value like our flow with the connectionString. We want the mask to be "generated" under certain criteria to produce `postgres://********asd` but we cannot apply these criteria to all values we received. As such it defaults to `************` unless an optional value is provided.

madebysid replied:
Ah that makes sense, we actually might want to use it right away, I like your idea of `postgresql://****`

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -8,68 +8,129 @@ import React, {
-      ref?: ForwardedRef<HTMLButtonElement>
-      getSecret?: () => string
+      isSecret?: false
+      permissionToReveal?: never
```
```suggestion
      canReveal?: never
```
?

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -8,68 +8,129 @@ import React, {
+    const [isLoading, setIsLoading] = useState(false)
+    const [queuedAction, setQueuedAction] = useState<Actions | null>(null)
+    const canToggleVisibility =
+      value && permissionToReveal ? true : !value && onReveal ? true : false
```
Same here, can we decompose this `? :` chain?

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -8,68 +8,129 @@ import React, {
-        setCopied(CopiedStatus.Success)
-      } catch (err) {
-        setCopied(CopiedStatus.Failed)
+      if (!value) {
```
IIUC there's an encoded assumption here that `value` will always be `null`/`undefined` when not revealed.
This is not always true since `value` is a controlled prop. It just happens to be, in the connection string use-case.

luanvdw replied:
This is just to cater for the use cases where a value is not provided upfront. Are you referring to the order of,

```
if (!value) {} else
```

as opposed to,

```
if (value) {} else
```

madebysid replied:
No that's what I mean, why would the value not be provided upfront? It is a controlled prop right (and a required one)

luanvdw replied:
With the ability to update the value on request (i.e. onReveal callback) we are open to scenarios where we might not have a provided value to being with for example where the API returned null for some value given some criteria and that value is the value that they would like to provide to the `<Code />`  component. Like in the case of connectionString.

This is to cater for that and unburden the usage of the component from needing to provide some sort of _temp_ valid input

madebysid replied:
Then it's the consumer's responsibility to take care of this, we shouldn't try to handle all this in a design system library. For errors etc, the consumer might want to show a toast or whatever, but it should be up to them to decide what to show (via the `value` prop). 

Keep this component as dumb as possible to encourage reuse  

If you take a look at my async onReveal, that works in this scenario as well, the consumer sets the value to `Fetching...` when onReveal is called, then sets it to the actual value when it has it. Of an error occurs, it might show a toast and show an error here (or flip the `canReveal` bool back to false), but the point is that that control needs to stay with the consumer.

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -8,68 +8,129 @@ import React, {
-        setCopied(CopiedStatus.Failed)
+      if (!value) {
+        if (onReveal) {
+          onReveal()
```
What if `onReveal` was allowed to be an async function (on top of a regular one)? Then you could do something like:

```typescript
const handleCodeCopy = async () => {
  await handleToggleVisibility()
  await navigator.clipboard.writeText(value)
  setCopied(CopiedStatus.Success)
}

const handleToggleVisibility() => async () => {
  setIsValueVisible(!isValueVisible)
  await onReveal?.() // `await` will do nothing if onReveal is not async
}
```

The expectation would be that the onReveal _could_ do async stuff, but the contract of a controlled `value` prop would still need to hold.

This would make it so our use-case in Cloud is supported, but for others, if this masked value is a frontend only thing, they can just set the `value` prop directly.

madebysid replied:
Lens:
```typescript
onReveal?: () => void | () => Promise<void>
```

Cloud:
```
const [value, setValue] = useState("")

<Code value={value} isSecret canReveal onReveal={async() => {
  setValue("Fetching connecting string...")
  try {
    await fetch()  
    setValue(fetchedValue)
  } catch(e) {
    toast()
    setValue("Unable")
  }
}}
/>
```

```
<Code value="secret" isSecret canReveal />
```

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -8,68 +8,129 @@ import React, {
-      } else setSecretValue("")
+    const renderValue = (): string => {
+      if (!value || !isValueVisible) {
+        if (isLoading) {
```
Instead of handling `isLoading` here, let's let the consumer handle it.

Not the biggest fan of my initial proposal to add an `isLoading` prop, but if we take my `async onReveal` suggestion, then we can let the outsider control what shows up (because they can set the `value` to something more contextually appropriate, like `Fetching your connection string... `

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -80,69 +141,77 @@ export const Code = forwardRef<HTMLButtonElement, CodeProps>(
+        }
+        setIsLoading(false)
+      }
+    }, [value])
```
I guess the `async onReveal` will also make it so this queuing logic wouldn't be required anymore.

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -80,69 +141,77 @@ export const Code = forwardRef<HTMLButtonElement, CodeProps>(
-          "relative",
-          "items-center rounded-lg",
+          "relative flex justify-start items-center",
+          "w-full px-4 rounded-lg",
```
If we're following the class "bucketing" logic, then `w-full` and `px-4` belong with `flex`, because they affect layout. `rounded-lg` affects appearance

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -80,69 +141,77 @@ export const Code = forwardRef<HTMLButtonElement, CodeProps>(
-              </pre>
-            </span>
-            <span className="absolute w-8 h-full right-0 top-0 bg-gradient-to-r from-transparent via-transparent to-gray-900" />
+        <span className="relative flex-grow overflow-hidden">
```
Is this `overflow-hidden` needed?

luanvdw replied:
Indeed, it provides a box model restriction.

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -80,69 +141,77 @@ export const Code = forwardRef<HTMLButtonElement, CodeProps>(
-            </span>
-            <span className="absolute w-8 h-full right-0 top-0 bg-gradient-to-r from-transparent via-transparent to-gray-900" />
+        <span className="relative flex-grow overflow-hidden">
+          <span className="flex py-3 max-w-full overflow-x-auto">
```
I noticed that when a scrollbar does appear, it messes up alignment of the "command" & the `prefix`, wouldn't that still be the case?

luanvdw replied:
Can you perhaps share a screenshot? 🙏  Just to make sure I understand correctly as the alignment on my side appears correct on the browsers I tested.

madebysid replied:
![Screenshot 2021-08-10 at 15 06 17](https://user-images.githubusercontent.com/7689783/128872749-879600a9-166e-4819-8e82-11ed2b6e09db.png)

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -80,69 +141,77 @@ export const Code = forwardRef<HTMLButtonElement, CodeProps>(
+            <pre className="font-mono whitespace-nowrap">{renderValue()}</pre>
           </span>
-        )}
+          <span className="absolute w-8 h-full right-0 top-0 bg-gradient-to-r from-transparent via-transparent to-gray-900" />
```
Let's add a comment here explaining that this is a "fade", because at first glace I wasn't sure what this was

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -80,69 +141,77 @@ export const Code = forwardRef<HTMLButtonElement, CodeProps>(
+                className={cn({
+                  "text-gray-200 hover:text-gray-300": canToggleVisibility,
+                  "text-gray-500": !canToggleVisibility,
+                })}
```
If we're adding text colors here, then we probably don't need them on the `div` on line 180.

OR

You could add styles on the `div` directly, and this should inherit them.

madebysid commented on src/components/code/Code.stories.tsx:
```diff
@@ -40,34 +30,121 @@ export default {
+ */
+export const WithoutValue = () => (
+  <Code prefix="$" value="" isSecret permissionToReveal />
+)
```
What use case does this cater to? I don't thik this component should care _what_ the `value` is, it's not its responsibility

luanvdw replied:
This is not for any intended / desired use case, it is to have a story showcasing the logic that even if a user has the necessary permission to reveal the string... if no value is provided then it will remain masked.

luanvdw commented:
@madebysid I have removed async / onReveal functionality completely.

Reasoning;

1. We do not have a use case for it currently.
2. Providing async functionality extends the possible paths of execution by a lot and catering for the various use cases within the same component, at the moment makes it messy (same reasoning as to why we extracted `InlineCode` to be its own component)
3. I will create a separate PR with a cleaned up async / onReveal logic or create it as a separate `AsyncCode` component

With this current status, the goal is to finalise this review and merge it so that the various use case on Cloud can be updated.
After which I will create a separate task for async / onReveal.

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -1,148 +1,139 @@
     const [copied, setCopied] = useState<CopiedStatus | null>(null)
-    const [secretValue, setSecretValue] = useState("")
-    const isSecretVisible = secretValue === ""
+    const canCopyValue = !isSecret || (isSecret && canReveal) ? true : false
```
I think you should be able to drop the ? true : false
```suggestion
    const canCopyValue = !isSecret || (isSecret && canReveal)
```

madebysid commented on src/components/code/Code.tsx:
```diff
@@ -1,148 +1,139 @@
+      if (isValueVisible) {
+        return value
+      } else {
+        return secretText
```
No need to have a `renderValue` function if it's just one conditional, let's inline it.

madebysid commented on src/components/code/Code.stories.tsx:
```diff
@@ -37,37 +44,33 @@ export default {
 }
 
-export const Standard = () => <Code prefix="$">git clone</Code>
+export const Standard = () => <Code value="git clone" />
```
Can we also add a story called WithPrefix?

madebysid commented on src/components/code/Code.stories.tsx:
```diff
@@ -37,37 +44,33 @@ export default {
-      yarn run build:package4
-    </Code>
-  )
-}
```
Re-add this WithLongText story?

madebysid reviewed (approved):
Makes sense! Pre-approving but left very minor comments
